src/Components/WebAssembly/WebAssembly/src/Services/DefaultWebAssemblyJSRuntime.cs

```diff
@@ -163,3 +170,6 @@ protected override Task TransmitStreamAsync(long streamId, DotNetStreamReference
         return TransmitDataStreamToJS.TransmitStreamAsync(this, "Blazor._internal.receiveWebAssemblyDotNetDataStream", streamId, dotNetStreamReference);
     }
 }
+
+[JsonSerializable(typeof(RootComponentOperationBatch))]
+internal sealed partial class DefaultWebAssemblyJSRuntimeJsonSerializerContext : JsonSerializerContext;
```

Comment:
I would strongly recommend creating one central `JsonSerializerContext` in your project that declares all needed types in one place. This should serve to improve both application size and startup time. Even though these might generate for different types superficially, transitively there is bound to be overlap (e.g. `JsonTypeInfo<int>`, `JsonTypeInfo<string>` but also shared internal types) and it would be best if these were only generated/initialized at a single location.

Reply:
I agree this would be ideal - Unfortunately, I haven't found a good way to achieve that in totality. The types that we serialize are split across multiple assemblies and shared sources. I fixed the example highlighted here, but I don't think there's much else we can easily improve. Maybe one possible improvement is having the `DefaultAntiforgeryStateProvider` shared source use a duck-typed `InternalJsonSerializerContext` that supports `AntiforgeryRequestToken`. So each project that references `DefaultAntiforgeryStateProvider` must also define an `InternalJsonSerializerContext`. But then we'd probably like to be sure that each `InternalJsonSerializerContext` has the same `JsonSourceGenerationOptions` configured... Hm..